lib/ecto/adapters/mysql/connection.ex

```diff
@@ -338,13 +338,10 @@ if Code.ensure_loaded?(Mariaex) do
       [name, ?. | quote_name(field)]
     end
 
-    defp expr({:&, _, [idx, fields, _counter]}, sources, query) do
-      {source, name, schema} = elem(sources, idx)
-      if is_nil(schema) and is_nil(fields) do
-        error!(query, "MySQL does not support selecting all fields from #{source} without a schema. " <>
-                      "Please specify a schema or specify exactly which fields you want to select")
-      end
-      intersperse_map(fields, ", ", &[name, ?. | quote_name(&1)])
+    defp expr({:&, _, [idx]}, sources, query) do
```

Comment:
Ecto 2.2 will no longer pass those weird and invalid `{:&, _, list_with_three_elements}`. It will pass `{:&, [], [counter]}` which should behave as the old `{:&, _, [counter, nil, nil]}`. If you want to remain backwards compatible, you just need to add this new clause and keep the old one around.